### DISCUSSION TITLE
GitHub / X 以外のアカウントとの紐づけ

### DISCUSSION THREAD

**Discussion by shun-shobon** [approvers/members-assoc#27] [Ideas]
**TL; DR**: members-assoc を Approvers の Auth0 的なものにしたい

現在 members-assoc では Discord の OAuth2.0 と API を使用して Approvers の各メンバーと GitHub / X のアカウント情報との紐づけを行っていますが，これに Discord のプロフィール取得 API 以外の機能(他サービスの OAuth2.0 など)を利用して，メンバーと他の任意のサービスとの紐づけを行える機能の追加を提案します．

これが実現されれば，メンバーと他サービスとの紐づけをシームレスに行うことができ，Bot 等の開発が楽になり，より文化の発展に貢献できると考えています．

自分としてはこの様なものを考えています．

- **members-assoc に新たな他サービスを連携させる**: members-assoc への登録が完了した後，追加で対応している他サービスとの連携を可能にする．(例: members-assoc でログイン(現状その様なものはない)した後に，WakaTime の OAuth2.0 を利用してメンバーと WakaTime の情報を紐付ける)
- **他サービスの API Token を利用できるようにする**: 上記の機能で他サービスの情報を紐づけた際，API Token を取得してそれを members-assoc にて保存， Bot などで利用できるようにする．
	- 上記の機能を実現するためには Bot への情報共有を可能にするための Token の払い出し機能や API Endpoint の追加が必要になると考えています．

**Comment by MikuroXina** [comment]
1 つ目ですが, この `members-assoc` では, できる限り各アカウントの識別情報 (ID やアカウント名) の関連付けを行うだけにとどめておきたいです. WakaTime のアカウント情報は良いですが, そのメトリクスを入れるような設計は考慮していません. 頻繁に更新してデータを整合させることもあまりやりたくないです.

2 つ目の各 API のプロキシになれる機能は面白そうです. members-assoc のトークンさえあれば, Twitter API などのトークンを自動更新しつつ返してくれるようにすれば便利そうです. 具体的な仕様はまた今度考えてみます.

**Reply by MikuroXina** [reply]
基本的な仕様はこのような感じでしょうか.

```mermaid
sequenceDiagram
    participant User
    participant UserApplication
    participant members-assoc
    participant R2
    participant ExternalAPI

    User->>UserApplication: Use some service
    UserApplication-->>User: Show OAuth auhtorization page
    User->>ExternalAPI: Authroize
    ExternalAPI->>UserApplication: Redirect and pass token
    UserApplication->>members-assoc: Register new service with token
    members-assoc-->>UserApplication: Give assoc-token

    Note over User,ExternalAPI: Some days later

    User->>UserApplication: Use some service
    UserApplication->>members-assoc: Pass assoc-token
    members-assoc->>members-assoc: Decode and verify assoc-token
    members-assoc->>R2: Get the associated token
    R2-->>members-assoc: Return the token
    members-assoc->>members-assoc: Check expiration time
    alt is expired
        members-assoc->>ExternalAPI: Refresh token
        ExternalAPI-->>members-assoc: Give new token
        members-assoc->>R2: Store the token
    end
    members-assoc-->>UserApplication: Give the token
    UserApplication-->>User: Give appropriate response
```

**Reply by shun-shobon** [reply]
> 1 つ目ですが, この members-assoc では, できる限り各アカウントの識別情報 (ID やアカウント名) の関連付けを行うだけにとどめておきたいです. WakaTime のアカウント情報は良いですが, そのメトリクスを入れるような設計は考慮していません. 頻繁に更新してデータを整合させることもあまりやりたくないです.

僕の言い方が悪かったんですが，WakaTime のメトリクスというよりかはアカウント情報(User ID など？)を members-assoc で保持しておきたいみたいな意図の発言でした．(書いてから思ったんですが WakaTime のアカウント情報だけ持ってて特に意味ないですね)

**Reply by shun-shobon** [reply]
> 基本的な仕様はこのような感じでしょうか.

自分が想定していたのはそんな感じです．図にして頂きありがとうございます！
